### PR TITLE
ft(ZENKO-1616): Update ObjectMD model for azure blob api

### PR DESCRIPTION
**Pull request by tmacro**
Add fields required for Azure Blob API support

### CODE REVIEW

**Review by nicolas2bert — APPROVED** [review]

File: lib/models/ObjectMD.js [hunk 1]
```diff
@@ -352,6 +356,46 @@ class ObjectMD {
...
+        this._data['creation-time'] = creationTime;
+        return this;
+    }
+
+    /**
+     * Returns Creation Date
+     *
+     * @return {string} Creation Date
+     */
+    getCreationTime() {
```
Comment (jonathan-gramain) [inline lib/models/ObjectMD.js:398]: Is there any expectation regarding compatibility with "old" metadata? I.e. do we need e.g. to fallback to `last-modified` if `creation-time` does not exist in the loaded metadata, or is it fine to return `undefined` in such case?

Comment (tmacro) [reply lib/models/ObjectMD.js:398]: Good question, I don't think this has been discussed

Comment (JianqinWang) [reply lib/models/ObjectMD.js:398]: In Mystique (R.I.P) we had `creation-time` fall back to `last-modified`

Comment (tmacro) [reply lib/models/ObjectMD.js:398]: I've added logic to fallback to `last-modified` if `creation-time` is not set

**Review by JianqinWang — APPROVED** [review]

**Comment by tmacro** [comment]
@bert-e approve

File: lib/models/ObjectMD.js [hunk 2]
```diff
@@ -910,12 +961,18 @@ class ObjectMD {
...
+     *                         (defaults to x-amz-meta- for backwards
+     *                          compatablility)
      * @return {ObjectMD} itself
      */
     setUserMetadata(metaHeaders) {
         Object.keys(metaHeaders).forEach(key => {
             if (key.startsWith('x-amz-meta-')) {
                 this._data[key] = metaHeaders[key];
+            } else if (key.startsWith('x-ms-meta-')) {
+                const _key = key.replace('x-ms-meta-', 'x-amz-meta-');
```
Comment (ploki) [inline lib/models/ObjectMD.js:971]: ~is this a rogue header ? ```x-ms-meta-x-ms-meta-plop```~
only the first occurrence is replaced. and the startsWidth() makes sure of this

File: lib/models/ObjectMD.js [hunk 3]
```diff
@@ -910,12 +961,18 @@ class ObjectMD {
      * Set custom meta headers
      *
      * @param {object} metaHeaders - Meta headers
+     * @param {sting} prefix - Prefix used to match meta headers
```
Comment (ploki) [inline lib/models/ObjectMD.js]: prefix no longer a parameter

File: lib/models/ObjectMD.js [hunk 4]
```diff
@@ -28,9 +28,14 @@ class ObjectMD {
             } else {
                 this._updateFromParsedJSON(objMd);
             }
+            if (!this.getCreationTime()) {
+                this.setCreationTime(this.getLastModified());
```
Comment (ploki) [inline lib/models/ObjectMD.js:32]: this block is never executed

**Review by ploki — CHANGES_REQUESTED** [review]

File: lib/models/ObjectMD.js [hunk 5]
```diff
@@ -352,6 +359,50 @@ class ObjectMD {
...
+    }
+
+    /**
+     * Returns Creation Date
+     *
+     * @return {string} Creation Date
+     */
+    getCreationTime() {
+        // If creation-time is not set fallback to LastModified
+        if (!this._data['creation-time']) {
```
Comment (ploki) [inline lib/models/ObjectMD.js:400]: From my understanding of the constructor code, this condition is never met, or if it is, last-modified is not defined either. Am I wrong?

**Review by ploki — APPROVED** [review]

File: lib/models/ObjectMD.js [hunk 6]
```diff
@@ -28,9 +28,14 @@ class ObjectMD {
...
                 this._updateFromParsedJSON(objMd);
             }
+            if (!this._data['creation-time']) {
+                this.setCreationTime(this.getLastModified());
+            }
         } else {
             // set newly-created object md modified time to current time
-            this._data['last-modified'] = new Date().toJSON();
+            const dt = new Date().toJSON();
+            this._data['last-modified'] = dt;
```
Comment (nicolas2bert) [inline lib/models/ObjectMD.js]: Is there anything that prevents us from using the setters?

**Review by rahulreddy — CHANGES_REQUESTED** [review]
Please increment Object Model version and document the attributes here https://github.com/scality/Arsenal/blob/development/8.1/lib/constants.js#L37

**Review by nicolas2bert — APPROVED** [review]

**Review by rahulreddy — APPROVED** [review]

**Pull request merged**